### PR TITLE
Convert postgres tables to citus local tables for ALTER TABLE commands defining foreign keys

### PR DESCRIPTION
**Pull request by onurctirtir**
As part of #4415, we need to intercept 
* `ALTER TABLE ADD FOREIGN KEY` and
* `ALTER TABLE ADD COLUMN REFERENCES`

commands involving postgres tables and reference (or citus local) tables to automatically convert postgres tables involved in this command -and the other postgres tables that those postgres tables are connected via their foreign key graphs- to citus local tables.

### CODE REVIEW

File: src/backend/distributed/commands/table.c [hunk 1]
```diff
@@ -593,59 +609,185 @@ PreprocessAlterTableStmt(Node *node, const char *alterTableCommand)
...
+ * GetAlterTableAddFKeyRightRelationIdList returns a list of oid's for right
+ * relations involved in foreign keys to be defined by given ALTER TABLE command.
+ */
+static List *
+GetAlterTableAddFKeyRightRelationIdList(AlterTableStmt *alterTableStatement)
+{
+	List *rightRelationRangeVarList =
+		GetAlterTableAddFKeyRightRelationRangeVarList(alterTableStatement);
+	List *commandList = alterTableStatement->cmds;
+	LOCKMODE lockMode = AlterTableGetLockLevel(commandList);
```
Comment (onderkalaci) [inline src/backend/distributed/commands/table.c:749]: I guess we should make this a little more explicit, iterating over `commandList ` and only acting on `FOREIGN KEY`s

File: src/backend/distributed/commands/table.c [hunk 2]
```diff
@@ -593,59 +609,185 @@ PreprocessAlterTableStmt(Node *node, const char *alterTableCommand)
...
+}
+
+
+/*
+ * ConvertPostgresLocalTablesToCitusLocalTables converts each postgres table
+ * involved in foreign keys to be defined by given alter table command and the
+ * other tables connected to them via a foreign key graph to citus local tables.
+ */
+static void
+ConvertPostgresLocalTablesToCitusLocalTables(AlterTableStmt *alterTableStatement)
```
Comment (onderkalaci) [inline src/backend/distributed/commands/table.c:662]: What about `alter table add column references`, should we handle that as well?

File: src/backend/distributed/commands/table.c [hunk 3]
```diff
@@ -593,59 +609,185 @@ PreprocessAlterTableStmt(Node *node, const char *alterTableCommand)
...
+			/*
+			 * relationRangeVarList has also reference and citus local tables
+			 * involved in this ADD FOREIGN KEY command. Morever, even if
+			 * rightRelationId was belonging to a postgres  local table initially,
+			 * we might had already converted it to a citus local table by cascading.
+			 */
+			continue;
 		}
+
+		/* TODO: should use try/catch block for CreateCitusLocalTable function */
```
Comment (onderkalaci) [inline src/backend/distributed/commands/table.c:705]: why do we need try catch?

File: src/backend/distributed/commands/table.c [hunk 4]
```diff
@@ -346,8 +345,15 @@ PostprocessAlterTableSchemaStmt(Node *node, const char *queryString)
  * function returns NIL.
  */
 List *
-PreprocessAlterTableStmt(Node *node, const char *alterTableCommand)
+PreprocessAlterTableStmt(Node *node, const char *alterTableCommand,
+						 ProcessUtilityContext processUtilityContext)
 {
+	if (processUtilityContext == PROCESS_UTILITY_SUBCOMMAND)
```
Comment (onderkalaci) [inline src/backend/distributed/commands/table.c:351]: can we move this check into / close to `AlterTableDefinesFKeyBetweenPostgresAndNonDistTable `?

File: src/backend/distributed/commands/table.c (hunk 3 above)
Comment (onurctirtir) [reply src/backend/distributed/commands/table.c:705]: to give little bit more precious error messages, like:

```c
try:
  CreateCitusLocalTable(cascade);
catch:
  ereport("Cannot define foreign key due to reason: %s", catchedMessage)
```

File: src/backend/distributed/commands/table.c (hunk 1 above)
Comment (onurctirtir) [reply src/backend/distributed/commands/table.c:749]: Indeed this is what we do by calling `GetAlterTableAddFKeyRightRelationRangeVarList -> GetAlterTableAddFKeyConstraintList -> GetAlterTableCommandFKeyConstraintList`

File: src/backend/distributed/commands/table.c (hunk 2 above)
Comment (onurctirtir) [reply src/backend/distributed/commands/table.c:662]: yes, we actually handle add column commands via `GetAlterTableCommandFKeyConstraintList` function

File: src/backend/distributed/commands/table.c [hunk 5]
```diff
@@ -376,14 +376,37 @@ PreprocessAlterTableStmt(Node *node, const char *alterTableCommand)
...
 
-	/*
-	 * Normally, we would do this check in ErrorIfUnsupportedForeignConstraintExists
-	 * in post process step. However, we skip doing error checks in post process if
-	 * this pre process returns NIL -and this method returns NIL if the left relation
-	 * is a postgres table. So, we need to error out for foreign keys from postgres
-	 * tables to citus local tables here.
-	 */
-	ErrorIfAlterTableDefinesFKeyFromPostgresToCitusLocalTable(alterTableStatement);
+	if (processUtilityContext == PROCESS_UTILITY_SUBCOMMAND)
```
Comment (onderkalaci) [inline src/backend/distributed/commands/table.c:379]: I was thinking something like the following where `processUtilityContext` is only relevant for `AlterTableDefinesFKeyBetweenPostgresAndNonDistTable`. Does the below work fine?
```C
	if (processUtilityContext != PROCESS_UTILITY_SUBCOMMAND && 
		AlterTableDefinesFKeyBetweenPostgresAndNonDistTable(alterTableStatement))
	{
```

File: src/backend/distributed/commands/table.c (hunk 1 above)
Comment (onderkalaci) [reply src/backend/distributed/commands/table.c:784]: I think what confuses me a lot is that we check the `AlterTableStmt` command type too late/too deep, inside `GetAlterTableCommandFKeyConstraintList `. Maybe in terms of correctness it could be OK, but readability/simplicity wise we could improve. 

Can we do something like the following, and use `l` in as a param to the other functions?
```C
static void
ConvertPostgresLocalTablesToCitusLocalTables(AlterTableStmt *alterTableStatement)
{
	List *l = GetAlterTableAddFKeyConstraintList(alterTableStatement);
	if (l == NIL)
		return;
```

Does that sound good to you?

File: src/backend/distributed/commands/table.c (hunk 2 above)
Comment (onderkalaci) [reply src/backend/distributed/commands/table.c:669]: See my other comment, I think I was confused because of that

File: src/test/regress/sql/fkeys_between_local_ref.sql [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+SELECT logicalrelid FROM pg_dist_partition, pg_tables
+WHERE tablename=logicalrelid::regclass::text AND
+      schemaname='fkeys_between_local_ref' AND
+      partmethod = 'n' AND repmodel = 'c';
+
+-- remove coordinator if it is added to pg_dist_node
+SELECT COUNT(master_remove_node(nodename, nodeport)) < 2
+FROM pg_dist_node WHERE nodename='localhost' AND nodeport=:master_port;
+
+CREATE TABLE local_table_1 (col_1 INT UNIQUE);
```
Comment (onderkalaci) [inline src/test/regress/sql/fkeys_between_local_ref.sql:22]: let's have some data inside the table

File: src/backend/distributed/commands/table.c [hunk 7]
```diff
@@ -593,59 +616,213 @@ PreprocessAlterTableStmt(Node *node, const char *alterTableCommand)
...
+			 * As we are in preprocess, missingOk might be true and relation
+			 * might not exist.
+			 */
+			continue;
 		}
+		else if (IsCitusTable(rightRelationId))
+		{
+			/*
+			 * relationRangeVarList has also reference and citus local tables
+			 * involved in this ADD FOREIGN KEY command. Morever, even if
```
Comment (onderkalaci) [inline src/backend/distributed/commands/table.c:705]: typo `Morever `

File: src/backend/distributed/commands/table.c [hunk 8]
```diff
@@ -593,59 +616,213 @@ PreprocessAlterTableStmt(Node *node, const char *alterTableCommand)
...
 
 	List *commandList = alterTableStatement->cmds;
 	AlterTableCmd *command = NULL;
 	foreach_ptr(command, commandList)
 	{
-		List *commandFKeyConstraintList = GetAlterTableCommandFKeyConstraintList(command);
-		alterTableFKeyConstraintList = list_concat(alterTableFKeyConstraintList,
-												   commandFKeyConstraintList);
+		List *commandForeignKeyConstraintList =
+			GetAlterTableCommandFKeyConstraintList(command);
```
Comment (onderkalaci) [inline src/backend/distributed/commands/table.c:820]: I think it safer to skip if `commandForeignKeyConstraintList  == NIL`

File: src/backend/distributed/commands/table.c [hunk 9]
```diff
@@ -593,59 +616,213 @@ PreprocessAlterTableStmt(Node *node, const char *alterTableCommand)
...
+ * reference or citus local table.
  */
-static void
-ErrorIfAlterTableDefinesFKeyFromPostgresToCitusLocalTable(
-	AlterTableStmt *alterTableStatement)
+static bool
+AlterTableDefinesFKeyBetweenPostgresAndNonDistTable(AlterTableStmt *alterTableStatement)
 {
-	List *commandList = alterTableStatement->cmds;
+	bool rightHasPostgresLocalTable = false;
```
Comment (onderkalaci) [inline src/backend/distributed/commands/table.c:626]: Following the 4 variables (`rightHasPostgresLocalTable`, `rightHasNonDistTable`, `leftIsNonDistTable`, `leftIsPostgresTable`) and their combinations are hard to follow.

Can we avoid variable names and come-up with something similar to the below?
```C

static bool
AlterTableDefinesFKeyBetweenPostgresAndNonDistTable(AlterTableStmt *alterTableStatement)
{
	bool rightHasPostgresLocalTable = false;
	bool rightHasNonDistTable = false;

	Oid rightRelationId = InvalidOid;
	List *rightRelationIdList =
		GetAlterTableAddFKeyRightRelationIdList(alterTableStatement);

	LOCKMODE lockmode = AlterTableGetLockLevel(alterTableStatement->cmds);
	Oid leftRelationId = AlterTableLookupRelation(alterTableStatement, lockmode);
	if (IsCitusTable(leftRelationId))
	{
		return RelationIdListContains(rightRelationIdList, CITUS_TABLE_WITH_NO_DIST_KEY);
	}
	else if (IsCitusTableType(leftRelationId, CITUS_TABLE_WITH_NO_DIST_KEY))
	{
		return !RelationIdListContains(rightRelationIdList, ANY_CITUS_TABLE_TYPE);
	}

	return false;
}
```

File: src/backend/distributed/commands/table.c [hunk 10]
```diff
@@ -593,59 +616,213 @@ PreprocessAlterTableStmt(Node *node, const char *alterTableCommand)
...
+			FlushErrorState();
+
+			if (errorData->elevel != ERROR)
+			{
+				/* don't try to intercept PANIC or FATAL */
+				PG_RE_THROW();
+			}
+
+			/* extend error message with foreign key creation error */
+			StringInfo errorMessage = makeStringInfo();
```
Comment (onderkalaci) [inline src/backend/distributed/commands/table.c:736]: Do we really need this try/catch block? We are conveying the same error message to the user with a  `cannot create foreign key` prefix, which doesn't sound crucial?

File: src/backend/distributed/commands/table.c [hunk 11]
```diff
@@ -593,59 +616,213 @@ PreprocessAlterTableStmt(Node *node, const char *alterTableCommand)
...
+	{
+		List *commandList = alterTableStatement->cmds;
+		LOCKMODE lockMode = AlterTableGetLockLevel(commandList);
+		bool missingOk = alterTableStatement->missing_ok;
+		Oid rightRelationId = RangeVarGetRelid(relationRangeVar, lockMode, missingOk);
+		if (!OidIsValid(rightRelationId))
 		{
-			ErrorOutForFKeyBetweenPostgresAndCitusLocalTable(leftRelationId);
+			/*
+			 * As we are in preprocess, missingOk might be true and relation
```
Comment (onderkalaci) [inline src/backend/distributed/commands/table.c:696]: Is this really possible? If so, what would be the consequence of skipping such relations?

File: src/backend/distributed/commands/table.c [hunk 12]
```diff
@@ -593,59 +616,213 @@ PreprocessAlterTableStmt(Node *node, const char *alterTableCommand)
...
+	 * Here we should operate on RangeVar objects since relations oid's would
+	 * change in below loop due to CreateCitusLocalTable.
+	 */
+	RangeVar *relationRangeVar;
+	foreach_ptr(relationRangeVar, relationRangeVarList)
+	{
+		List *commandList = alterTableStatement->cmds;
+		LOCKMODE lockMode = AlterTableGetLockLevel(commandList);
+		bool missingOk = alterTableStatement->missing_ok;
+		Oid rightRelationId = RangeVarGetRelid(relationRangeVar, lockMode, missingOk);
```
Comment (onderkalaci) [inline src/backend/distributed/commands/table.c:692]: minor: we should probably not use the name `rightRelationId ` as the left relation is also on the list

File: src/backend/distributed/commands/table.c (hunk 8 above)
Comment (onurctirtir) [reply src/backend/distributed/commands/table.c:820]: afaik `list_concat` already handles empty lists, so do you think we need extra check here ?

Comment (onderkalaci) [reply src/backend/distributed/commands/table.c:820]: Yes, it'd be cleaner

File: src/backend/distributed/commands/table.c (hunk 5 above)
Comment (onurctirtir) [reply src/backend/distributed/commands/table.c:379]: probably yes, let me try

Comment (onurctirtir) [reply src/backend/distributed/commands/table.c]: 17bc6100bc8d66a7c834babb10bba25fcbfc3609

File: src/backend/distributed/commands/table.c (hunk 7 above)
Comment (onurctirtir) [reply src/backend/distributed/commands/table.c]: d3ab13514e97bb0964c2121d1710adfa0230b5e9

File: src/backend/distributed/commands/table.c (hunk 12 above)
Comment (onurctirtir) [reply src/backend/distributed/commands/table.c]: f242d116e61aa05d66421bae5ee667c62bf4f5f2

File: src/backend/distributed/commands/table.c (hunk 9 above)
Comment (onurctirtir) [reply src/backend/distributed/commands/table.c:620]: what about this ? (to be more explicit for postgres tables in `else if`)

```c
static bool
AlterTableDefinesFKeyBetweenPostgresAndNonDistTable(AlterTableStmt *alterTableStatement)
{
	List *rightRelationIdList =
		GetAlterTableAddFKeyRightRelationIdList(alterTableStatement);

	LOCKMODE lockmode = AlterTableGetLockLevel(alterTableStatement->cmds);
	Oid leftRelationId = AlterTableLookupRelation(alterTableStatement, lockmode);
	if (!IsCitusTable(leftRelationId))
	{
		return RelationIdListContainsCitusTableType(rightRelationIdList, CITUS_TABLE_WITH_NO_DIST_KEY);
	}

	if (IsCitusTableType(leftRelationId, CITUS_TABLE_WITH_NO_DIST_KEY))
	{
		return RelationIdListContainsPostgresTable(rightRelationIdList);
	}

	return false;
}
```

Comment (onurctirtir) [reply src/backend/distributed/commands/table.c]: e75d64d40d77bd98ab5574eb23ab57b376b5e4f2

File: src/backend/distributed/commands/table.c (hunk 11 above)
Comment (onurctirtir) [reply src/backend/distributed/commands/table.c:714]: Actually we don't expect invalid relations as `alter table` (afaik) has no such syntax for `add foreign key/column references`.
But as c struct has this field, I thought that it might be better to not relying on that.

>If so, what would be the consequence of skipping such relations?

If the thing I wrote above is true, i.e if no such syntax exists, `standard_ProcessUtility would error out.
Else, we will just skip invalid relations and I think we should be fine.

File: src/backend/distributed/commands/table.c (hunk 1 above)
Comment (onurctirtir) [reply src/backend/distributed/commands/table.c:809]: I'm okay with doing this check beforehand:

>	List *l = GetAlterTableAddFKeyConstraintList(alterTableStatement);
	if (l == NIL)
		return false;

but passing this list to below calls might decrease the readability I think, what do you think ?

File: src/backend/distributed/commands/table.c [hunk 13]
```diff
@@ -593,59 +613,243 @@ PreprocessAlterTableStmt(Node *node, const char *alterTableCommand)
...
+		{
+			MemoryContextSwitchTo(savedMemoryContext);
+
+			ErrorData *errorData = CopyErrorData();
+			FlushErrorState();
+
+			if (errorData->elevel != ERROR)
+			{
+				/* don't try to intercept PANIC or FATAL */
+				PG_RE_THROW();
```
Comment (marcocitus) [inline src/backend/distributed/commands/table.c:759]: I don't think PANIC or FATAL ever land in a PG_CATCH

File: src/backend/distributed/commands/table.c [hunk 14]
```diff
@@ -593,59 +613,243 @@ PreprocessAlterTableStmt(Node *node, const char *alterTableCommand)
...
+	List *rightRelationRangeVarList =
+		GetAlterTableAddFKeyRightRelationRangeVarList(alterTableStatement);
+	RangeVar *leftRelationRangeVar = alterTableStatement->relation;
+	List *relationRangeVarList = lappend(rightRelationRangeVarList, leftRelationRangeVar);
+
+	/*
+	 * To prevent deadlocks, sort the list before converting each postgres local
+	 * table to a citus local table.
+	 */
+	relationRangeVarList = SortList(relationRangeVarList, CompareRangeVarsByOid);
```
Comment (marcocitus) [inline src/backend/distributed/commands/table.c:707]: That does not really prevents deadlocks, but I suppose it can't hurt.

File: src/test/regress/expected/multi_foreign_key.out [hunk 15]
```diff
@@ -920,9 +920,7 @@ SELECT create_reference_table('reference_table');
 (1 row)
 
 ALTER TABLE reference_table ADD CONSTRAINT fk FOREIGN KEY(referencing_column) REFERENCES referenced_local_table(id);
-ERROR:  referenced table "referenced_local_table" must be a distributed table or a reference table
-DETAIL:  To enforce foreign keys, the referencing and referenced rows need to be stored on the same node.
-HINT:  You could use SELECT create_reference_table('referenced_local_table') to replicate the referenced table to all nodes
+ERROR:  cannot create foreign key: could not find the coordinator node in metadata as it is not added as a worker
```
Comment (marcocitus) [inline src/test/regress/expected/multi_foreign_key.out:923]: "as it is node added as a worker" seems a little confusing

maybe omit that part and add an:
errhint("Use SELECT citus_set_coordinator_host(...) to add the coordinator to the metadata")

File: src/backend/distributed/commands/table.c [hunk 16]
```diff
@@ -376,14 +379,31 @@ PreprocessAlterTableStmt(Node *node, const char *alterTableCommand)
...
-	 * Normally, we would do this check in ErrorIfUnsupportedForeignConstraintExists
-	 * in post process step. However, we skip doing error checks in post process if
-	 * this pre process returns NIL -and this method returns NIL if the left relation
-	 * is a postgres table. So, we need to error out for foreign keys from postgres
-	 * tables to citus local tables here.
-	 */
-	ErrorIfAlterTableDefinesFKeyFromPostgresToCitusLocalTable(alterTableStatement);
+	if (processUtilityContext != PROCESS_UTILITY_SUBCOMMAND &&
+		AlterTableDefinesFKeyBetweenPostgresAndNonDistTable(alterTableStatement))
+	{
```
Comment (marcocitus) [inline src/backend/distributed/commands/table.c:384]: The following steps work on 9.5/master, but fail on this branch when the coordinator is not added to the metadata
```sql
create table ref (a int primary key);
select create_reference_table('ref');
create table other (x int primary key, y int);
alter table other add constraint fk foreign key (y) references ref (a) on delete cascade;
select create_reference_table('other');
```

```sql
create table ref (a int primary key);
select create_reference_table('ref');
create table other (x int primary key, y int);
alter table other add constraint fk foreign key (y) references ref (a) on delete cascade;
ERROR:  cannot create foreign key: could not find the coordinator node in metadata as it is not added as a worker
```

What happens if we skip this if when the coordinator is not in metadata?

File: src/backend/distributed/commands/table.c (hunk 13 above)
Comment (onurctirtir) [reply src/backend/distributed/commands/table.c:759]: makes sense, then keeping `PG_RE_THROW();` for debug like messages but removing comment

File: src/backend/distributed/commands/table.c (hunk 14 above)
Comment (onurctirtir) [reply src/backend/distributed/commands/table.c:707]: true, it can't prevent many cases but at least could prevent same ddl command from different sessions

File: src/test/regress/expected/multi_foreign_key.out (hunk 15 above)
Comment (onurctirtir) [reply src/test/regress/expected/multi_foreign_key.out:923]: omitting error part only for `as it is node added as a worker` would probably be a little hard, do you think should we hide error part in try/catch block in `table.c`, like that:

```
try:
    CreateCitusLocalTable(cascadeViaForeignKeys)
catch:
    ereport(err="cannot create foreign key", hint=capturedHint)
```

File: src/backend/distributed/commands/table.c (hunk 16 above)
Comment (onderkalaci) [reply src/backend/distributed/commands/table.c:384]: With the changes on #4489, the above seems to be covered. Still, the following is not:
```SQL
drop table if exists ref, ref2 cascade;


create table ref (a int primary key);
create table ref2 (x int);
alter table ref2 add constraint fk foreign key (x) references ref (a);
select create_reference_table('ref');
alter table ref2 add constraint fk2 foreign key (x) references ref (a);

ERROR:  cannot create foreign key: table "ref" is already distributed
Time: 2.276 ms
 ```
It looks like `create_reference_table` does not convert `ref2` to a citus local table, and then the second `ALTER TABLE` cascades from `ref2` to `ref` for creating citus local tables as where already exists a foreign key between them.

In an ideal world, we'd want `create_reference_table` to cascade referencing tables to local tables. However, given that we may skip cascade if the coordinator is not in the metadata, we could consider `CascadeOperationForConnectedRelations` to skip cascading into already `IsCitusTable` relations when the operation is `create_citus_local`

Comment (onurctirtir) [reply src/backend/distributed/commands/table.c:384]: I think we have two issues to fix here:

For the first `alter table ref2 add constraint` command, we should be converting postgres tables to citus local tables as we do in this pr.

For the second `alter table ref2 add constraint`, as @onderkalaci mentioned, we need to skip citus tables when calling `CreateCitusLocalTable`.

This is because if we only fix the second part, then even if the coordinator is added to metadata, then we won't be converting local tables to citus local tables for such `create_reference_table` calls, but seems that we should do that.

File: src/test/regress/expected/multi_foreign_key.out (hunk 15 above)
Comment (marcocitus) [reply src/test/regress/expected/multi_foreign_key.out:923]: well I think we should adjust the original error, but let's not do this now

File: src/test/regress/expected/fkeys_between_local_ref.out [hunk 17]
```diff
@@ -0,0 +1,144 @@
...
+     tablename     | partmethod | repmodel
+---------------------------------------------------------------------
+ local_table_1     | n          | c
+ local_table_2     | n          | c
+ local_table_3     | n          | c
+ local_table_4     | n          | c
+ reference_table_1 | n          | t
+ reference_table_2 | n          | t
+(6 rows)
+
```
Comment (onurctirtir) [inline src/test/regress/expected/fkeys_between_local_ref.out:141]: noting here not to forget:

let's add some tests for unsupported fkeys between citus local tables & reference tables (like ON UPDATE CASCADE)

File: src/backend/distributed/commands/table.c [hunk 18]
```diff
@@ -376,14 +379,38 @@ PreprocessAlterTableStmt(Node *node, const char *alterTableCommand)
...
-	 * Normally, we would do this check in ErrorIfUnsupportedForeignConstraintExists
-	 * in post process step. However, we skip doing error checks in post process if
-	 * this pre process returns NIL -and this method returns NIL if the left relation
-	 * is a postgres table. So, we need to error out for foreign keys from postgres
-	 * tables to citus local tables here.
-	 */
-	ErrorIfAlterTableDefinesFKeyFromPostgresToCitusLocalTable(alterTableStatement);
+	if (processUtilityContext != PROCESS_UTILITY_SUBCOMMAND &&
+		AlterTableDefinesFKeyBetweenPostgresAndNonDistTable(alterTableStatement) &&
+		CoordinatorAddedAsWorkerNode())
```
Comment (onurctirtir) [inline src/backend/distributed/commands/table.c:384]: add some tests when coordinator not added as a worker

Comment (onderkalaci) [reply src/backend/distributed/commands/table.c:384]: yes, we need that

File: src/backend/distributed/commands/table.c [hunk 19]
```diff
@@ -690,6 +900,47 @@ GetAlterTableCommandFKeyConstraintList(AlterTableCmd *command)
 }
 
 
+/*
+ * GetRangeVarListFromConstraintList returns a list of RangeVar objects for
+ * right relations in constraintList.
+ */
+static List *
+GetRangeVarListFromConstraintList(List *constraintList)
```
Comment (onderkalaci) [inline src/backend/distributed/commands/table.c:908]: This name looks too generic: `GetRangeVarListFromForeignConstraintList `

File: src/test/regress/expected/fkeys_between_local_ref.out (hunk 17 above)
Comment (onderkalaci) [reply src/test/regress/expected/fkeys_between_local_ref.out:142]: yes, let's have one test like that (or data doesn't good to create fkey)

File: src/backend/distributed/commands/cascade_table_operation_for_connected_relations.c [hunk 20]
```diff
@@ -345,6 +345,13 @@ ExecuteCascadeOperationForRelationIdList(List *relationIdList,
 		CascadeOperationFunction cascadeOperationFunction =
 			GetCascadeOperationFunction(cascadeOperationType);
 
+		if (cascadeOperationType == CREATE_CITUS_LOCAL_TABLE &&
+			IsCitusTable(relationId))
+		{
```
[the remark this answers is not in the document]
Comment (onurctirtir) [reply src/backend/distributed/commands/cascade_table_operation_for_connected_relations.c:355]: After rebase it looks event much strange than it is, I will look into that

File: src/backend/distributed/commands/table.c (hunk 16 above)
Comment (onderkalaci) [reply src/backend/distributed/commands/table.c:385]: I'm resolving this as  all the mentioned items are covered for both the coordinator added or not.

File: src/backend/distributed/commands/table.c [hunk 21]
```diff
@@ -593,59 +620,242 @@ PreprocessAlterTableStmt(Node *node, const char *alterTableCommand)
...
+			/*
+			 * relationRangeVarList has also reference and citus local tables
+			 * involved in this ADD FOREIGN KEY command. Moreover, even if
+			 * relationId was belonging to a postgres  local table initially,
+			 * we might had already converted it to a citus local table by cascading.
+			 */
+			continue;
+		}
+
+		/*
```
Comment (onderkalaci) [inline src/backend/distributed/commands/table.c:746]: lets mention that we have this try/catch only for better error messages

File: src/test/regress/expected/fkeys_between_local_ref.out [hunk 22]
```diff
@@ -0,0 +1,145 @@
...
+ALTER TABLE local_table_4 ADD CONSTRAINT fkey_6 FOREIGN KEY (col_1) REFERENCES local_table_4(col_1);
+CREATE TABLE reference_table_1(col_1 INT UNIQUE, col_2 INT);
+INSERT INTO reference_table_1 SELECT i FROM generate_series(195, 205) i;
+SELECT create_reference_table('reference_table_1');
+ create_reference_table
+---------------------------------------------------------------------
+
+(1 row)
+
+-- we fall back to legacy behavior when coordinator is not added to metadata
```
Comment (onderkalaci) [inline src/test/regress/expected/fkeys_between_local_ref.out:50]: the comment and the behavior don't match. I think it is expected that `ALTER TABLE` fails.

I think we can consider adding the examples we discussed on the PR for coordinator not added case:
```SQL


create table ref (a int primary key);
select create_reference_table('ref');
create table other (x int primary key, y int);
alter table other add constraint fk foreign key (y) references ref (a) on delete cascade;
select create_reference_table('other');

create table ref (a int primary key);
select create_reference_table('ref');
create table other (x int primary key, y int);
alter table other add constraint fk foreign key (y) references ref (a) on delete cascade;


drop table if exists ref, ref2 cascade;


create table ref (a int primary key);
create table ref2 (x int);
alter table ref2 add constraint fk foreign key (x) references ref (a);
select create_reference_table('ref');
alter table ref2 add constraint fk2 foreign key (x) references ref (a);

ERROR:  cannot create foreign key: table "ref" is already distributed
Time: 2.276 ms
```

File: src/test/regress/expected/fkeys_between_local_ref.out (hunk 17 above)
Comment (onderkalaci) [reply src/test/regress/expected/fkeys_between_local_ref.out:142]: ping on this

File: src/backend/distributed/commands/table.c [hunk 23]
```diff
@@ -593,59 +620,242 @@ PreprocessAlterTableStmt(Node *node, const char *alterTableCommand)
...
+
+			if (errorData->elevel != ERROR)
+			{
+				PG_RE_THROW();
+			}
+
+			/* extend error message with foreign key creation error */
+			StringInfo errorMessage = makeStringInfo();
+			appendStringInfo(errorMessage, "cannot create foreign key: %s",
+							 errorData->message);
```
Comment (onurctirtir) [inline src/backend/distributed/commands/table.c:771]: maybe get rid of as we are getting weird outputs like that:

>ERROR:  cannot create foreign key: cannot create foreign key constraint since foreign keys ....

File: src/backend/distributed/commands/table.c (hunk 18 above)
Comment (onderkalaci) [reply src/backend/distributed/commands/table.c:384]: we added it but still requires some clean-up. Sill, resolving this as we discuss there

File: src/backend/distributed/commands/table.c (hunk 23 above)
Comment (onderkalaci) [reply src/backend/distributed/commands/table.c:771]: I think we could improve it as:
```
Creating foreign key failed to cascade because: %s
```

File: src/test/regress/expected/fkeys_between_local_ref.out (hunk 17 above)
Comment (onurctirtir) [reply src/test/regress/expected/fkeys_between_local_ref.out:178]: 14b8b5e6ba48ed367d3aff15fd66cdec2c4bafd5

File: src/backend/distributed/commands/cascade_table_operation_for_connected_relations.c (hunk 20 above)
[the remark this answers is not in the document]
Comment (onurctirtir) [reply src/backend/distributed/commands/cascade_table_operation_for_connected_relations.c]: 2e4b2ab12088ca29e5ffd9933f0e1ea38f7cb505

File: src/backend/distributed/commands/table.c (hunk 19 above)
Comment (onurctirtir) [reply src/backend/distributed/commands/table.c]: 2014e976980bb84d1bbe74218bb2ca7403ec9689

File: src/backend/distributed/commands/table.c (hunk 21 above)
Comment (onurctirtir) [reply src/backend/distributed/commands/table.c:746]: cd5647b56a86dc2a592bbaa3df92e5f73536d431

File: src/test/regress/expected/fkeys_between_local_ref.out (hunk 22 above)
Comment (onurctirtir) [reply src/test/regress/expected/fkeys_between_local_ref.out]: 0bac4344012c9cd19356b6aff2ef6dd0a0540690

**Review by onderkalaci — APPROVED** [review]
